src/include/storage/index/bplustree.h

```diff
@@ -0,0 +1,2554 @@
+          return false;
+        }
+        itr++;
+      }
+      // current node checked, move to the next one
+      node = reinterpret_cast<ElasticNode<KeyValuePair> *>(node->GetLowKeyPair().second);
+    }
+    return true;
+  }
+
+  /**
+   * This test checks if duplicate keys are correctly handled by the insert wrapper.
+   * For a B+ tree, with each key having multiple values, stored in a map keys_values
+   * check if the values_list in the tree have the same values as the map given
+   * in the function.
+   *
+   * NOTE: This function does not acquire any latches, used by tests
+   */
+  bool DuplicateKeyValuesCheck(std::unordered_map<KeyType, std::set<ValueType>> *keys_values) {
+    auto itr = (*keys_values).begin();
+    for (; itr != (*keys_values).end(); itr++) {
+      KeyType k = itr->first;
+      std::set<ValueType> values = (*keys_values)[k];
+      std::vector<ValueType> result;
+      FindValueOfKey(k, &result);
+      if (result.empty()) {
+        return false;
+      }
+      bool return_value = true;
+      for (auto it = result.begin(); it != result.end(); it++) {
+        if (!(return_value)) {
+          break;
+        }
+        if (values.count(*(it)) == 0) {
+          return_value = false;
+        } else {
+          values.erase(*(it));
+        }
+      }
+      if (!(return_value)) {
+        return return_value;
+      }
+      if (!values.empty()) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * The check gets a map of keys to vector of values that are whetted and
+   * removes any duplicates that might have been tried to be added by a
+   * transaction in insert unique. The check ensures that the values list
+   * in the B+ tree also have the same keys and the same order.
+   *
+   * NOTE: This function does not acquire any latches, used by tests
+   */
+  bool DuplicateKeyValueUniqueInsertCheck(std::unordered_map<KeyType, std::vector<ValueType>> *keys_values) {
+    auto itr = (*keys_values).begin();
+    for (; itr != (*keys_values).end(); itr++) {
+      KeyType k = itr->first;
+      std::vector<ValueType> values = (*keys_values)[k];
+      std::list<ValueType> *values_list_p = FindValueOfKey(k);
+      if (values_list_p == nullptr) {
+        return false;
+      }
+      auto it_list = values_list_p->begin();
+      for (unsigned j = 0; j != values.size(); j++) {
+        if (it_list == values_list_p->end()) {
+          return false;
+        }
+        if (values[j] != *(it_list)) {
+          return false;
+        }
+        it_list++;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Returns current heap usage (tries to return the exact heap usage by traversing
+   * the tree).
+   *
+   * NOTE: This function acquires a coarse lock, on the root, and prevents access to
+   * any other thread. Not ideal behavior, but this function is not commonly used. To
+   * get an approximate heap usage, EstimateHeapUsage() should be used.
+   */
+  size_t GetHeapUsage() {
```

Comment:
Is there a reason that a coarse lock is used?

Reply:
This function returns the **exact** heap usage. So technically, if we use fine-grained locking, the measurement will not be accurate since the tree can change dynamically. This function sort of captures the current snapshot. For estimates of heap usage, EstimateHeapUsage() function should be called.